docs: add common chart configuration options (#3353)
akuity/kargo: docs/docs/60-new-docs/40-operator-guide/20-advanced-installation/30-common-configurations.md

configurations for Kargo. 

:::info
For complete parameter documentation, refer to the
[chart documentation](https://github.com/akuity/kargo/blob/main/charts/kargo/README.md).
:::

:::info
For more information on how to apply these configurations, see the
advanced installation guides for [Helm](10-advanced-with-helm.md) and
[Argo CD](20-advanced-with-argocd.md).
:::

## Standard Kubernetes Configuration

For common Kubernetes configurations, like setting a `nodeSelector`, `labels`,
`annotations`, `affinity`, `tolerations`, etc. Kargo supports both global and
resource-specific configurations. For example, to add `labels` to all resources
created by Kargo, the following configuration can be used:

```yaml
global:
  labels:
    label.example.com/key: value
```

To add `labels` to a specific component (the `kargo-controller` in this example),
the following configuration can be used:
    
```yaml
controller:
  labels:
    label.example.com/key: value
```

:::note
For a full list of supported configurations, refer to the
[Global Parameters](https://github.com/akuity/kargo/blob/main/charts/kargo/README.md#global-parameters)

> or the component-specific parameters in the chart documentation.

Reviewer: ```suggestion
or the component-specific parameters sections in the chart documentation.
```

:::

## API Configuration

Kargo supports a number of API-related configurations that can be set at
installation time. These configurations are used to control the behavior of
Kargo's API server and its web-based user interface.

> :::info
> The sections below outline common configurations for the API server. For a full
> list of supported configurations, refer to the
> [API Parameters](](https://github.com/akuity/kargo/blob/main/charts/kargo/README.md#api).

Reviewer: ```suggestion
:::info
The sections below outline common configurations for the API server. For a full
list of supported configurations, refer to the
[API Parameters](https://github.com/akuity/kargo/blob/main/charts/kargo/README.md#api)
section in the chart documentation.
:::
```

### API Host

By default, the API server host (i.e. the domain or IP address that the API is
accessible at) is set to `localhost`. This host is used for generation of

> Ingress resources, certificates, the OpenID Connect issuer and callback URLs,

Reviewer: ```suggestion
`Ingress` resources, certificates, the OpenID Connect issuer and callback URLs,
```

and any URLs that are exposed to users.

To configure the API host, set the following configuration:

```yaml
api:
  host: kargo.example.com
```

:::note
The host is allowed to include a port number, e.g. `kargo.example.com:8080`,
but should not include a protocol (e.g. `http://` or `https://`) as this is
automatically inferred from other configuration options.
:::

### API Service

By default, Kargo will create a `Service` resource for the API server with the
type `ClusterIP`, which is only accessible within the cluster or through
[port forwarding](https://kubernetes.io/docs/tasks/access-application-cluster/port-forward-access-application-cluster/).

:::caution
Changing the API server service type can expose the API server to the internet
in an insecure manner. Refer to the
[Secure Configuration](../40-security/10-secure-configuration.md) section for
more information on securing the API server.
:::

:::info
Instead of making use of a `Service` resource, you can also expose the API

> server using an [Ingress resource](#api-ingress).

Reviewer: I could go either way on monospacing technical terms within links, but I've seen you do this in several places, so I'm following you lead. If you disagree, that's fine too.

```suggestion
server using an [`Ingress` resource](#api-ingress).
```

:::

If you want to expose the API server to the internet, but do not want to make
use of an [Ingress resource](#api-ingress), you can change the service type:

```yaml
api:
  service:
    type: LoadBalancer
```

In addition, when using a `LoadBalancer` or `NodePort` service type, you can
configure the service to use a specific port:

```yaml
api:
  service:
    type: LoadBalancer
    port: 443
```

### API TLS

By default, Kargo will enable TLS directly on the API server using a
self-signed certificate issued by [cert-manager](https://cert-manager.io/).

:::note
When making use of the self-signed certificate option, cert-manager must be
installed in the cluster.
:::

To supply your own certificate, set the following configuration:

```yaml
api:
  tls:
    selfSignedCert: false
```

When setting `selfSignedCert` to `false`, a
[TLS `Secret`](https://kubernetes.io/docs/concepts/configuration/secret/#tls-secrets)
with the name `kargo-api-cert` is expected to exist in the same namespace as
the API server.

#### Terminating TLS

In certain cases, you may want to disable TLS on the API server because it is
terminated at the [`Ingress`](#api-ingress) level. To do this, set the following

> configuration:

Reviewer: There's some nuance to this that isn't captured here. If you opt-in to Ingress + TLS through the chart, everything just works.

The `terminatedUpstream` option is needed when something other than that has terminated TLS upstream. It could be a reverse proxy, a load balancer, an Ingress controller that terminates TLS using its _own_ wildcard cert instead of a cert associated directly with the Ingress resource, or it could even be terminated by some other form of routing like the Gateway API (which we should probably support sooner or later) or a Traefik IngressRoute.

So `terminatedUpstream` is an escape hatch for any scenario where generating links should use https, but Kargo has no other way of inferring that from its own configuration.

more information on securing the API server.
:::

To configure the API server to use an `Ingress` resource, set the following
configuration:

```yaml
api:
  ingress:
    enabled: true
    ingressClassName: nginx
```

#### Ingress TLS

By default, Kargo will enable TLS on the `Ingress` resource using a self-signed
certificate using [cert-manager](https://cert-manager.io). 

:::note
When making use of the self-signed certificate option, cert-manager must be
installed in the cluster.
:::

To supply your own certificate, set the following configuration:

```yaml
api:
  ingress:
    tls:
      selfSignedCert: false
```

When setting `selfSignedCert` to `false`, the `Ingress` resource expects a
[TLS `Secret`](https://kubernetes.io/docs/concepts/services-networking/ingress/#tls)
with the name `kargo-api-ingress-cert` to exist in the same namespace as the
API server.

## Git Configuration

Kargo supports a number of Git-related configurations that can be set at
installation time.

### Default Commit Author

To set the default commit author Kargo will use when committing changes using
the [`git-commit` Promotion step](../../50-user-guide/60-reference-docs/30-promotion-steps/git-commit.md),

> the following configuration can be set (shown here with the default values):

Reviewer: Is it worth mentioning that users can override this default in their own configurations? Or leave that the the step docs you're already linking to? I could go either way.

Author: The intention of the link was to point people towards the overwrite.

Reviewer: Fair enough. I suppose the word "default" also implies is well enough that it's overridable, even if the reader doesn't follow the link to the git-commit step.

should be a GPG key in ASCII-armored format  without a passphrase, under the
key `signingKey`.

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: kargo-git-signing-key
type: Opaque
data:
  signingKey: <base64-encoded-ascii-armored-gpg-key>
```

The `Secret` can then be referenced in the Kargo configuration.

```yaml
controller:
  gitClient:
    signingKeySecret:
      name: kargo-git-signing-key
      type: gpg
```

:::note
When using a signing key, the `gitClient.name` and `gitClient.email`
configuration options must match the name and email associated with the GPG
key.
:::

## Argo CD Configuration

Kargo supports a number of Argo CD-related configurations that can be set at
installation time.

### Disabling the Argo CD Integration

By default, Kargo will enable the Argo CD integration, which configures Kargo
to work with `Application` resources created by Argo CD. This can be disabled
as follows:

```yaml
controller:
  argocd:
    integrationEnabled: false
```

When not enabled, the controller will not watch Argo CD `Application` resources
and disable Argo CD specific features. Explicitly disabling is preferred if this

> integration is not desired, as it will grant fewer permissions to the controller.

Reviewer: So we want to mention in this paragraph the behavior of degrading to a disabled state if the controller learns at startup that Argo CD CRDs are not present in the cluster?

I think that segues well into the point you're already making about explicitly disabling being preferable if the capability is not wanted.

### Argo CD Namespace

By default, Kargo expects Argo CD to be installed to the `argocd` namespace,
which is also the default namespace it will use for `Application` resources
if a namespace is not specified in the
[`argocd-update` Promotion step](../../50-user-guide/60-reference-docs/30-promotion-steps/argocd-update.md).

If you want to use a different default namespace, this can be configured as
follows:

```yaml
controller:
  argocd:
    namespace: argocd
```

In certain cases, you may not want Kargo to be able to look up Argo CD resources
in other namespaces than the configured default. This can be enforced by setting

> the following configuration:

Reviewer: Ha. I forgot that the default for this was false. It's an _opt-in_. Thanks for refreshing my memory. 😄

```yaml
controller:
  argocd:
    watchArgocdNamespaceOnly: true
```

When enabled, the controller will only watch for `Application` resources in the
namespace specified by `controller.argocd.namespace`.

## Argo Rollouts Configuration

### Disabling the Argo Rollouts Integration

By default, Kargo will enable the Argo Rollouts integration, which configures
Kargo to work with `Rollout` resources created by Argo Rollouts as part of the

> [verification feature](../../50-user-guide/20-how-to-guides/14-working-with-stages.md#verifications).

Reviewer: Love that you mention what we _do_ use it for, but only because I'm aware of it having been a recent point of confusion for a few users, we probably need to be explicit that Kargo does not in any way require that your own application configurations utilize Rollouts.

Reviewer: Fair enough. @jessesuen has that in-progress and I am sure he will mention that. We'll just update this link to the improved doc when it's complete.

This can be disabled as follows:

```yaml
controller:
  argoRollouts:
    integrationEnabled: false
```

When not enabled, the controller will not reconcile Argo Rollouts `AnalysisRun`
resources and attempts to verify Stages via `Analysis` will fail. Explicitly
disabling is preferred if this integration is not desired, as it will grant

> fewer permissions to the controller.

Reviewer: Similar question to [here](https://github.com/akuity/kargo/pull/3353/files#r1928887163)

## Resource Management

### Tuning Concurrent Reconciliation Limits

By default, Kargo will reconcile up to 4 resources of the same kind concurrently
(e.g. 4 `Stage` resources at a time). It is possible to tune this limit by

> setting a new global value, or by setting a limit per resource kind:

Reviewer: Nit: Not that we follow any one of these in particular, but all of the common style guides (used in the US, at least) agree that 4 should be "four" here.

https://www.masterclass.com/articles/when-to-spell-out-numbers-explained

Not a hill I would die on by any means.

[verification feature](../../50-user-guide/20-how-to-guides/14-working-with-stages.md#verifications).

This can be disabled as follows:

```yaml
controller:
  argoRollouts:
    integrationEnabled: false
```

When not enabled, the controller will not reconcile Argo Rollouts `AnalysisRun`
resources and attempts to verify Stages via `Analysis` will fail. Explicitly
disabling is preferred if this integration is not desired, as it will grant
fewer permissions to the controller.

## Resource Management

### Tuning Concurrent Reconciliation Limits

By default, Kargo will reconcile up to 4 resources of the same kind concurrently
(e.g. 4 `Stage` resources at a time). It is possible to tune this limit by
setting a new global value, or by setting a limit per resource kind:

```yaml
controller:
  reconcilers:
    # Global setting
    maxConcurrentReconciles: 4

  managementController:
    # Kind specific setting
    namespaces:
      maxConcurrentReconciles: 2
```

:::note
For a list of resource kinds that can be configured, refer to the
[chart documentation](https://github.com/akuity/kargo/blob/main/charts/kargo/README.md).
:::

> ## Garbage Collection

Reviewer: I would have forgotten about this. Glad you remembered.

more information on securing the API server.
:::

:::info
Instead of making use of a `Service` resource, you can also expose the API
server using an [`Ingress` resource](#api-ingress).
:::

If you want to expose the API server to the internet, but do not want to make
use of an `Ingress` resource, you can change the service type:

```yaml
api:
  service:
    type: LoadBalancer
```

In addition, when using a `LoadBalancer` or `NodePort` service type, you can
configure the service to use a specific port:

```yaml
api:
  service:
    type: LoadBalancer
    port: 443
```

### API TLS

By default, Kargo will enable TLS directly on the API server using a
self-signed certificate issued by [cert-manager](https://cert-manager.io/).

:::note
When making use of the self-signed certificate option, cert-manager must be
installed in the cluster.
:::

To supply your own certificate, set the following configuration:

```yaml
api:
  tls:
    selfSignedCert: false
```

When setting `selfSignedCert` to `false`, a
[TLS `Secret`](https://kubernetes.io/docs/concepts/configuration/secret/#tls-secrets)
with the name `kargo-api-cert` is expected to exist in the same namespace as
the API server.

> #### Terminating TLS
> 
> In some cases, TLS may be terminated upstream of the API server by various
> components like reverse proxies, load balancers, or `Ingress` controllers using
> wildcard certificates.
> 
> For these cases, the API server can be configured to not enforce TLS:
> 
> ```yaml
> api:
>   tls:
>     enabled: false
>     terminatedUpstream: true
> ```
> 
> :::info
> When setting `terminatedUpstream` to `true`, the API server will generate URLs
> with the `https` scheme but will not enforce TLS. This option serves as an
> escape hatch for scenarios where HTTPS should be used for links, but Kargo
> cannot infer this from its own configuration.
> 
> When you are using the [built-in `Ingress` resource](#api-ingress), this
> configuration is typically not required.
> :::

Reviewer: I think there was some good stuff in the info box at the bottom that needs a brighter spotlight on it. Can I suggest this, where the bit about Kargo being unable to infer TLS is in use is moved up higher?

Absolutely feel free to tweak further.

```suggestion
#### Terminating TLS Upstream

By default, Kargo will enable TLS on the API server itself and the `Ingress` resource (if it's enabled) using self-signed certificates issued by [cert-manager](https://cert-manager.io). 

If you have opted out of these and TLS is terminated upstream from the API server
by a component such as a reverse proxy, load balancer, or `Ingress` controller using
a wildcard certificate, Kargo will be unable to infer from its own configuration
that any generated URLs should use HTTPS.

For such cases, you can explicitly declare that TLS is terminated upstream and Kargo will sort out the rest:

```yaml
api:
  tls:
    enabled: false
    terminatedUpstream: true
```

:::info
If you exercise any of the chart's built-in TLS configuration options,
setting `terminatedUpstream` is never required.
:::
```

Reviewer: Note: I re-tweaked my suggestion.

Author: Thank you, applied.
